File: backend/app/schemas/survey.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class SurveyGenreCreate(BaseModel):
    genre_id: int = Field(validation_alias="id")
    is_liked: bool


class SurveyGenreRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int = Field(validation_alias="genre_id")
    is_liked: bool


class SurveyCreate(BaseModel):
    genres: list[SurveyGenreCreate]
    animes: list[int]


class SurveyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    user_id: int
    genres: list[SurveyGenreRead]
    animes: list[int]
    created_at: datetime
    updated_at: datetime

    @field_validator("animes", mode="before")
    @classmethod
    def _animes_to_ids(cls, v) -> list[int]:
        return [a.id for a in v]


class SurveyUpdate(SurveyCreate):
    pass

File: frontend/types/survey.ts
export interface SurveyGenrePayload {
  id: number
  is_liked: boolean
}

export interface SurveyPayload {
  genres: SurveyGenrePayload[]
  animes: number[]
}

export interface SurveyRead extends SurveyPayload {
  id: number
  user_id: number
  created_at: string
  updated_at: string
}

export interface SurveyGenreFormData {
  id: number
  isLiked: boolean
}

export interface SurveyFormData {
  genres: SurveyGenreFormData[]
  animes: number[]
}

export interface Survey extends SurveyFormData {
  id: number
  userId: number
  createdAt: string
  updatedAt: string
}
